src/main/java/com/conveyal/datatools/manager/controllers/api/LabelController.java

```diff
@@ -0,0 +1,224 @@
+package com.conveyal.datatools.manager.controllers.api;
+
+import com.conveyal.datatools.manager.auth.Actions;
+import com.conveyal.datatools.manager.auth.Auth0UserProfile;
+import com.conveyal.datatools.manager.models.FeedSource;
+import com.conveyal.datatools.manager.models.JsonViews;
+import com.conveyal.datatools.manager.models.Label;
+import com.conveyal.datatools.manager.models.Project;
+import com.conveyal.datatools.manager.persistence.Persistence;
+import com.conveyal.datatools.manager.utils.json.JsonManager;
+import org.apache.commons.lang3.StringUtils;
+import org.eclipse.jetty.http.HttpStatus;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import spark.Request;
+import spark.Response;
+
+import java.io.IOException;
+import java.util.*;
+import java.util.stream.Collectors;
+
+import static com.conveyal.datatools.common.utils.SparkUtils.getPOJOFromRequestBody;
+import static com.conveyal.datatools.common.utils.SparkUtils.logMessageAndHalt;
+import static spark.Spark.*;
+import static spark.Spark.post;
+
+
+public class LabelController {
+    private static final Logger LOG = LoggerFactory.getLogger(LabelController.class);
+    private static JsonManager<Label> json = new JsonManager<>(Label.class, JsonViews.UserInterface.class);
+
+
+    /**
+     * Spark HTTP endpoint to get a single label by ID
+     */
+    private static Label getLabel(Request req, Response res) {
+        return requestLabelById(req, Actions.VIEW);
+    }
+
+    /**
+     * Spark HTTP endpoint that handles getting all labels for a handful of use cases:
+     * - for a single project (if projectId query param provided)
+     * - for the entire application
+     */
+    private static Collection<Label> getAllLabels(Request req, Response res) {
+        Collection<Label> labelsToReturn = new ArrayList<>();
+        Auth0UserProfile user = req.attribute("user");
+        String projectId = req.queryParams("projectId");
+        Project project = Persistence.projects.getById(projectId);
+        if (project == null) {
+            logMessageAndHalt(req, 400, "Must provide valid projectId query param to retrieve labels.");
+        }
+
+        Collection<Label> projectLabels = project.retrieveProjectLabels();
+        for (Label label: projectLabels) {
+            // Only return labels user has access to.
+            labelsToReturn.add(checkLabelPermissions(req, label, Actions.VIEW));
```

Comment:
This could end up inserting null labels. Please refactor so that only non-null values are returned.

Reply:
Actually, this current design would end up halting the request instead of returning non-admin-viewable labels.

Reply:
This is a good catch thank you! Resolved in 
85261da